orttraining/orttraining/core/optimizer/triton_fusion.cc

```diff
@@ -0,0 +1,432 @@
+    if (axis < 0) axis += rank;
+    int64_t non_neg_expected_axis = expected_axis < 0 ? expected_axis + rank : expected_axis;
+    return axis == non_neg_expected_axis;
+  }
+  return false;
+}
+
+bool CheckAxes(const Graph& graph, const Node& node, bool single_axis, const std::vector<int>& expected_axes) {
+  std::vector<int64_t> axes_values;
+  const auto& attributes = node.GetAttributes();
+  if (attributes.find("axes") != attributes.end()) {
+    axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+  } else if (node.InputDefs().size() == 2) {
+    auto axes_const = graph.GetConstantInitializer(node.InputDefs()[1]->Name(), true);
+    if (!axes_const) {
+      return false;
+    }
+    Initializer initializer{*axes_const, graph.ModelPath()};
+    axes_values.insert(axes_values.end(), initializer.DataAsSpan<int64_t>().begin(),
+                       initializer.DataAsSpan<int64_t>().end());
+  } else {
+    return false;
+  }
+
+  if (expected_axes.empty()) {
+    return !single_axis || axes_values.size() == 1;
+  }
+
+  std::vector<int64_t> expected_axes_values(expected_axes.begin(), expected_axes.end());
+  bool has_negative_axis = false;
+  for (auto axis : axes_values) {
+    if (axis < 0) {
+      has_negative_axis = true;
+      break;
+    }
+  }
+  if (!has_negative_axis) {
+    for (auto axis : expected_axes_values) {
+      if (axis < 0) {
+        has_negative_axis = true;
+        break;
+      }
+    }
+  }
+  if (has_negative_axis) {
+    const auto& input_shape = node.InputDefs()[0]->Shape();
+    if (input_shape == nullptr) return false;
+    int64_t rank = static_cast<int64_t>(input_shape->dim_size());
+    for (auto& axis : axes_values) {
+      if (axis < 0) axis += rank;
+    }
+    for (auto& axis : expected_axes_values) {
+      if (axis < 0) axis += rank;
+    }
+  }
+  std::sort(axes_values.begin(), axes_values.end());
+  std::sort(expected_axes_values.begin(), expected_axes_values.end());
+  return axes_values == expected_axes_values;
+}
+
+struct Partition {
+  NodeVec nodes;
+  NodeArgSet outputs;
+  NodeArgSet dependencies;
+  size_t output_ref_count;
+
+  void MergeFrom(const Partition& other) {
+    nodes.insert(nodes.end(), other.nodes.begin(), other.nodes.end());
+    outputs.insert(other.outputs.begin(), other.outputs.end());
+    dependencies.insert(other.dependencies.begin(), other.dependencies.end());
+    output_ref_count += other.output_ref_count;
+  }
+
+  bool IsValid(const TritonFusionConfig& config) const {
+    size_t count = 0;
+    bool all_ignore_min_nodes = true;
+    for (const auto& node : nodes) {
+      if (!config.IsNoOp(*node)) {
+        ++count;
+        if (count >= config.min_nodes) return true;
+      }
+      if (!config.IgnoreMinNodes(*node)) all_ignore_min_nodes = false;
+    }
+    return all_ignore_min_nodes;
+  }
+};
+
+}  // namespace
+
+void from_json(const json& j, TritonFusionConfig::OpInfo& op_info) {
```

Comment:
to FromJson?

Reply:
It's conventional name from Json lib. https://github.com/nlohmann/json